🛠️ Task 1: Data Setup & Validation
Use your knowledge of the os module and DataFrame attributes to prepare the environment.

Check Existence: Use os.path.exists() to verify the CSV is in your folder before loading.

Inspect Structure: Print the .info() and .describe() to see the average age and price.

Missing Values: Check if there are any null values using .isnull().sum().

In [12]:
import os
import pandas as pd
import numpy as np
import random
import datetime as dt

In [13]:
# read the csv file, hence the file exists

df = pd.read_csv('retail_sales_dataset.csv')
df.head()

Transaction ID        Date Customer ID  Gender  Age Product Category  \
0               1  2023-11-24     CUST001    Male   34           Beauty   
1               2  2023-02-27     CUST002  Female   26         Clothing   
2               3  2023-01-13     CUST003    Male   50      Electronics   
3               4  2023-05-21     CUST004    Male   37         Clothing   
4               5  2023-05-06     CUST005    Male   30           Beauty   

   Quantity  Price per Unit  Total Amount  
0         3              50           150  
1         2             500          1000  
2         1              30            30  
3         1             500           500  
4         2              50           100

In [14]:
# basic info about the data

print(df.info())
print('============================================================================')
print('')
print('============================================================================')
print(df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    1000 non-null   int64 
 1   Date              1000 non-null   object
 2   Customer ID       1000 non-null   object
 3   Gender            1000 non-null   object
 4   Age               1000 non-null   int64 
 5   Product Category  1000 non-null   object
 6   Quantity          1000 non-null   int64 
 7   Price per Unit    1000 non-null   int64 
 8   Total Amount      1000 non-null   int64 
dtypes: int64(5), object(4)
memory usage: 70.4+ KB
None

       Transaction ID         Age     Quantity  Price per Unit  Total Amount
count     1000.000000  1000.00000  1000.000000     1000.000000   1000.000000
mean       500.500000    41.39200     2.514000      179.890000    456.000000
std        288.819436    13.68143     1.132734      189.681356    559.997632
min          1.000000   

In [15]:
# checking for null values

df.isnull().sum()

Transaction ID      0
Date                0
Customer ID         0
Gender              0
Age                 0
Product Category    0
Quantity            0
Price per Unit      0
Total Amount        0
dtype: int64

🛠️ Task 2: Time-Based Feature Engineering
The Date column is currently just text (an object). Let's make it useful:

Convert to Datetime: Use pd.to_datetime(df['Date']).

Extract Month: Create a new column Month by extracting the month from the date.

Question: Which month in 2023 had the highest total sales?


In [16]:
df.head()

Transaction ID        Date Customer ID  Gender  Age Product Category  \
0               1  2023-11-24     CUST001    Male   34           Beauty   
1               2  2023-02-27     CUST002  Female   26         Clothing   
2               3  2023-01-13     CUST003    Male   50      Electronics   
3               4  2023-05-21     CUST004    Male   37         Clothing   
4               5  2023-05-06     CUST005    Male   30           Beauty   

   Quantity  Price per Unit  Total Amount  
0         3              50           150  
1         2             500          1000  
2         1              30            30  
3         1             500           500  
4         2              50           100

In [17]:
# convert date object from an object to a 'datetime'

df['Date'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    1000 non-null   int64         
 1   Date              1000 non-null   datetime64[ns]
 2   Customer ID       1000 non-null   object        
 3   Gender            1000 non-null   object        
 4   Age               1000 non-null   int64         
 5   Product Category  1000 non-null   object        
 6   Quantity          1000 non-null   int64         
 7   Price per Unit    1000 non-null   int64         
 8   Total Amount      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(5), object(3)
memory usage: 70.4+ KB


In [18]:
# extract the month from the Date and create a new column

df.insert(2, 'Month', df['Date'].dt.month)
df.head()

Transaction ID       Date  Month Customer ID  Gender  Age Product Category  \
0               1 2023-11-24     11     CUST001    Male   34           Beauty   
1               2 2023-02-27      2     CUST002  Female   26         Clothing   
2               3 2023-01-13      1     CUST003    Male   50      Electronics   
3               4 2023-05-21      5     CUST004    Male   37         Clothing   
4               5 2023-05-06      5     CUST005    Male   30           Beauty   

   Quantity  Price per Unit  Total Amount  
0         3              50           150  
1         2             500          1000  
2         1              30            30  
3         1             500           500  
4         2              50           100

In [ ]:
# check the month with the highest total sales

df.value_counts('Month')


Month
5     105
10     96
8      94
12     91
4      86
2      85
1      78
11     78
6      77
3      73
7      72
9      65
Name: count, dtype: int64

### Question 1: Which month in 2023 had the highest total sales?

### Answer: the month with the highest number of sales was Month 5 (May)

🛠️ Task 3: Customer Demographic Analysis
This dataset includes Gender and Age, which is perfect for understanding who is buying.

Gender Split: Use .value_counts() on the Gender column to see the breakdown of shoppers.

Age Bins: Create a simple function and use .apply() to categorize customers into Young (under 30), Adult (30-50), and Senior (50+).

Question: Which age group spends the most on average?

In [20]:
df.head()

Transaction ID       Date  Month Customer ID  Gender  Age Product Category  \
0               1 2023-11-24     11     CUST001    Male   34           Beauty   
1               2 2023-02-27      2     CUST002  Female   26         Clothing   
2               3 2023-01-13      1     CUST003    Male   50      Electronics   
3               4 2023-05-21      5     CUST004    Male   37         Clothing   
4               5 2023-05-06      5     CUST005    Male   30           Beauty   

   Quantity  Price per Unit  Total Amount  
0         3              50           150  
1         2             500          1000  
2         1              30            30  
3         1             500           500  
4         2              50           100

In [21]:
df.value_counts('Gender')

Gender
Female    510
Male      490
Name: count, dtype: int64

In [39]:
# 1. Create an empty list to store our categories
age_groups = []

# 2. Loop through the Age column in your dataset
for age in df['Age']:
    if age < 30:
        age_groups.append('Young')
    elif age <= 50:
        age_groups.append('Adult')
    else:
        age_groups.append('Senior')


# 3. Add the list as a new column in your DataFrame
df['Age Group'] = age_groups

df

Transaction ID       Date  Month Customer ID  Gender  Age  \
0                 1 2023-11-24     11     CUST001    Male   34   
1                 2 2023-02-27      2     CUST002  Female   26   
2                 3 2023-01-13      1     CUST003    Male   50   
3                 4 2023-05-21      5     CUST004    Male   37   
4                 5 2023-05-06      5     CUST005    Male   30   
..              ...        ...    ...         ...     ...  ...   
995             996 2023-05-16      5     CUST996    Male   62   
996             997 2023-11-17     11     CUST997    Male   52   
997             998 2023-10-29     10     CUST998  Female   23   
998             999 2023-12-05     12     CUST999  Female   36   
999            1000 2023-04-12      4    CUST1000    Male   47   

    Product Category  Quantity  Price per Unit  Total Amount Age Group  
0             Beauty         3              50           150     Adult  
1           Clothing         2             500          1000     Young  
2        Electronics         1              30            30     Adult  
3           Clothing         1             500           500     Adult  
4             Beauty         2              50           100     Adult  
..               ...       ...             ...           ...       ...  
995         Clothing         1              50            50    Senior  
996           Beauty         3              30            90    Senior  
997           Beauty         4              25           100     Young  
998      Electronics         3              50           150     Adult  
999      Electronics         4              30           120     Adult  

[1000 rows x 11 columns]

In [ ]:
# which age group spend the most on average?

df.groupby('Age Group')['Total Amount'].sum()


Age Group
Adult     199535
Senior    133310
Young     123155
Name: Total Amount, dtype: int64

### Question 2: Which age group spends the most on average?

### Answer: The Adults (ages 30 - 50) spent the most on average ($199,535)

🛠️ Task 4: Product Category Performance
This is where your groupby and sort_values skills shine.

Top Categories: Group the data by Product Category and calculate the sum of Total Amount.

Ranking: Use .sort_values() to rank categories from highest revenue to lowest.

Question: Does "Electronics" generate more revenue than "Clothing"?

In [51]:
df.head()

Transaction ID       Date  Month Customer ID  Gender  Age Product Category  \
0               1 2023-11-24     11     CUST001    Male   34           Beauty   
1               2 2023-02-27      2     CUST002  Female   26         Clothing   
2               3 2023-01-13      1     CUST003    Male   50      Electronics   
3               4 2023-05-21      5     CUST004    Male   37         Clothing   
4               5 2023-05-06      5     CUST005    Male   30           Beauty   

   Quantity  Price per Unit  Total Amount Age Group  
0         3              50           150     Adult  
1         2             500          1000     Young  
2         1              30            30     Adult  
3         1             500           500     Adult  
4         2              50           100     Adult

In [57]:
# group data by product category

df.groupby('Product Category')['Total Amount'].sum().sort_values(ascending=False)

Product Category
Electronics    156905
Clothing       155580
Beauty         143515
Name: Total Amount, dtype: int64

Question 3: Does "Electronics" generate more revenue than "Clothing"?

Answer: Yes "Electronics" generates more revenue than "Clothing" ($156905 > $155580)